pkg/services/dashboards/service/zanzana.go

```diff
@@ -0,0 +1,325 @@
+package service
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/grafana/pkg/services/accesscontrol"
+	"github.com/grafana/grafana/pkg/services/authz/zanzana"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	"github.com/grafana/grafana/pkg/services/sqlstore/searchstore"
+)
+
+const (
+	maxListQueryLength = 8
+	maxListQueryLimit  = 50
+	defaultQueryLimit  = 1000
+)
+
+type searchResult struct {
+	runner   string
+	result   []dashboards.DashboardSearchProjection
+	err      error
+	duration time.Duration
+}
+
+func (dr *DashboardServiceImpl) FindDashboardsZanzana(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery) ([]dashboards.DashboardSearchProjection, error) {
+	if dr.cfg.Zanzana.SingleRead {
+		return dr.findDashboardsZanzanaSingleRead(ctx, query)
+	}
+	return dr.findDashboardsZanzanaCompare(ctx, query)
+}
+
+func (dr *DashboardServiceImpl) findDashboardsZanzanaSingleRead(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery) ([]dashboards.DashboardSearchProjection, error) {
+	timer := prometheus.NewTimer(dr.metrics.searchRequestsDuration.WithLabelValues("zanzana"))
+	defer timer.ObserveDuration()
+
+	return dr.findDashboardsZanzana(ctx, query)
+}
+
+func (dr *DashboardServiceImpl) findDashboardsZanzanaCompare(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery) ([]dashboards.DashboardSearchProjection, error) {
+	result := make(chan searchResult, 2)
+
+	go func() {
+		timer := prometheus.NewTimer(dr.metrics.searchRequestsDuration.WithLabelValues("zanzana"))
+		defer timer.ObserveDuration()
+		start := time.Now()
+
+		queryZanzana := *query
+		res, err := dr.findDashboardsZanzana(ctx, &queryZanzana)
+		result <- searchResult{"zanzana", res, err, time.Since(start)}
+	}()
+
+	go func() {
+		timer := prometheus.NewTimer(dr.metrics.searchRequestsDuration.WithLabelValues("grafana"))
+		defer timer.ObserveDuration()
+		start := time.Now()
+
+		res, err := dr.FindDashboards(ctx, query)
+		result <- searchResult{"grafana", res, err, time.Since(start)}
+	}()
+
+	first, second := <-result, <-result
+	close(result)
+
+	if second.runner == "grafana" {
+		first, second = second, first
+	}
+
+	if second.err != nil {
+		dr.log.Error("zanzana search failed", "error", second.err)
+	} else if len(first.result) != len(second.result) {
+		dr.metrics.searchRequestStatusTotal.WithLabelValues("error").Inc()
+		dr.log.Warn(
+			"zanzana search result does not match grafana",
+			"grafana_result_len", len(first.result),
+			"zanana_result_len", len(second.result),
+			"grafana_duration", first.duration,
+			"zanzana_duration", second.duration,
+		)
+	} else {
+		dr.metrics.searchRequestStatusTotal.WithLabelValues("success").Inc()
+		dr.log.Debug("zanzana search is correct", "result_len", len(first.result), "grafana_duration", first.duration, "zanzana_duration", second.duration)
```

Comment:
Yes, currently we want to use simple indicator like list length to catch major issues with a search. Full correctness is hard to achieve because of ordering. If number of search results is higher than limit, then order might not be preserved in zanzana queries and results are not match. But in this case I believe it's not a big issue, since usually you should narrow the search to get less results which is easy to handle.